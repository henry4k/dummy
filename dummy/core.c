#include <stdlib.h> // malloc, free
#include <string.h> // memset, strncpy
#include <stdio.h>  // dummyFormat etc
#include <stdarg.h> // dummyFormat etc
#include <stdbool.h>
#include <assert.h>

#include "core.h"


enum
{
    DUMMY_MAX_MESSAGE_LENGTH = 127,
    DUMMY_MAX_TESTS = 32,
    DUMMY_MAX_CLEANUPS = 32,
    DUMMY_INVALID_TEST_INDEX = -1,
    DUMMY_SANDBOX_SKIPPED = -1
};

typedef enum
{
    DUMMY_INITIALIZING,
    DUMMY_RUNNING,
    DUMMY_COMPLETED
} dummyStatus;

typedef enum
{
    DUMMY_TEST_UNDEFINED,
    DUMMY_TEST_STARTING,
    DUMMY_TEST_RUNNING,
    DUMMY_TEST_COMPLETED
} dummyTestStatus;

/**
 * Is called after the test has been completed -
 * regardless of whether it failed or succeeded.
 */
typedef struct
{
    dummyCleanupFunction fn;
    void* data;
} dummyCleanup;

typedef struct
{
    char name[DUMMY_MAX_MESSAGE_LENGTH];
    dummySandbox sandbox;
    dummySandboxableFunction fn;

    dummyTestStatus status;

    dummyTestResult result;
    char abortReason[DUMMY_MAX_MESSAGE_LENGTH];

    bool markedAsTodo;
    char todoReason[DUMMY_MAX_MESSAGE_LENGTH];

    dummyCleanup cleanupStack[DUMMY_MAX_CLEANUPS];
    int cleanupStackSize;
} dummyTest;

typedef struct
{
    dummyStatus status;
    const dummyReporter* reporter;

    dummyTest tests[DUMMY_MAX_TESTS];
    int testCount;
    int currentTestIndex;
} dummyContext;

typedef struct
{
    dummyAbortHandler handler;
    void* context;
} dummyAbortHandlerInfo;


static dummyContext* dummyCurrentContext = NULL;
static dummyAbortHandlerInfo AbortStack[DUMMY_MAX_SANDBOX_DEPTH];
static int AbortStackSize = 0;


bool dummyRunTest( int index );
dummyTest* dummyGetCurrentTest();

void dummyInit( const dummyReporter* reporter )
{
    assert(dummyCurrentContext == NULL);
    dummyCurrentContext = (dummyContext*)malloc(sizeof(dummyContext));
    memset(dummyCurrentContext, 0, sizeof(dummyContext));

    dummyCurrentContext->status = DUMMY_INITIALIZING;

    assert(reporter);
    assert(reporter->began);
    assert(reporter->completed);
    assert(reporter->beganTest);
    assert(reporter->completedTest);
    assert(reporter->log);
    dummyCurrentContext->reporter = reporter;

    dummyCurrentContext->currentTestIndex = DUMMY_INVALID_TEST_INDEX;
}

int dummyRunTests()
{
    dummyContext* ctx = dummyCurrentContext;
    assert(ctx);
    assert(ctx->status == DUMMY_INITIALIZING);

    ctx->reporter->began(ctx->reporter->context);

    ctx->status = DUMMY_RUNNING;
    int failedTests = 0;
    int i = 0;
    for(; i < ctx->testCount; i++)
        if(dummyRunTest(i) == false)
            failedTests++;

    ctx->status = DUMMY_COMPLETED;
    ctx->reporter->completed(ctx->reporter->context);

    assert(ctx->status == DUMMY_COMPLETED);
    free(ctx);
    dummyCurrentContext = NULL;

    return failedTests;
}

void dummyAddTest( const char* name, dummySandbox sandbox, dummySandboxableFunction fn )
{
    dummyContext* ctx = dummyCurrentContext;
    assert(ctx);
    assert(ctx->status == DUMMY_INITIALIZING);
    assert(ctx->testCount < DUMMY_MAX_TESTS);

    dummyTest* test = &ctx->tests[ctx->testCount];
    memset(test, 0, sizeof(dummyTest));

    strncpy(test->name, name, DUMMY_MAX_MESSAGE_LENGTH);
    test->sandbox = sandbox;
    test->fn = fn;

    ctx->testCount++;
}

bool dummyRunTest( int index )
{
    dummyContext* ctx = dummyCurrentContext;
    assert(ctx);
    assert(ctx->status == DUMMY_RUNNING);

    assert(index >= 0);
    assert(index < DUMMY_MAX_TESTS);
    dummyTest* test = &ctx->tests[index];

    // prepare
    ctx->currentTestIndex = index;
    test->status = DUMMY_TEST_STARTING;
    test->result = DUMMY_TEST_PASSED;
    ctx->reporter->beganTest(ctx->reporter->context);

    // run
    test->status = DUMMY_TEST_RUNNING;
    const char* abortReason = NULL;
    const int errorCode = test->sandbox(test->fn, &abortReason);
    switch(errorCode)
    {
        case DUMMY_SANDBOX_SUCCEEDED:
            test->result = DUMMY_TEST_PASSED;
            break;

        case DUMMY_SANDBOX_SKIPPED:
            test->result = DUMMY_TEST_SKIPPED;
            break;

        case DUMMY_SANDBOX_GENERIC_ERROR:
        default:
            test->result = DUMMY_TEST_FAILED;
    }
    if(abortReason)
        strncpy(test->abortReason, abortReason, DUMMY_MAX_MESSAGE_LENGTH);
    test->status = DUMMY_TEST_COMPLETED;

    for(int i = 0; i < test->cleanupStackSize; i++)
    {
        dummyCleanup* cleanup = &test->cleanupStack[i];
        cleanup->fn(cleanup->data);
    }
    test->cleanupStackSize = 0;

    // cleanup
    ctx->reporter->completedTest(ctx->reporter->context);
    ctx->currentTestIndex = DUMMY_INVALID_TEST_INDEX;
    test->status = DUMMY_TEST_UNDEFINED;

    return test->result != DUMMY_TEST_FAILED;
}

int dummyGetTestCount()
{
    const dummyContext* ctx = dummyCurrentContext;
    assert(ctx);

    return ctx->testCount;
}

int dummyGetTestNumber()
{
    const dummyContext* ctx = dummyCurrentContext;
    assert(ctx);

    assert(ctx->currentTestIndex >= 0);
    assert(ctx->currentTestIndex < DUMMY_MAX_TESTS);

    return ctx->currentTestIndex;
}

dummyTest* dummyGetCurrentTest()
{
    dummyContext* ctx = dummyCurrentContext;
    assert(ctx);

    assert(ctx->currentTestIndex >= 0);
    assert(ctx->currentTestIndex < DUMMY_MAX_TESTS);
    dummyTest* test = &ctx->tests[ctx->currentTestIndex];

    return test;
}

const char* dummyGetTestName()
{
    const dummyTest* test = dummyGetCurrentTest();
    assert(test->name);
    assert(strlen(test->name) < DUMMY_MAX_MESSAGE_LENGTH);
    return test->name;
}

dummyTestResult dummyGetTestResult()
{
    const dummyTest* test = dummyGetCurrentTest();
    assert(test->status == DUMMY_TEST_COMPLETED);
    return test->result;
}

const char* dummyGetTestAbortReason()
{
    const dummyTest* test = dummyGetCurrentTest();
    if(test->status == DUMMY_TEST_COMPLETED)
    {
        assert(test->abortReason);
        const int length = strlen(test->abortReason);
        assert(length < DUMMY_MAX_MESSAGE_LENGTH);
        if(length > 0)
            return test->abortReason;
    }

    return NULL;
}

void dummyAddCleanup( dummyCleanupFunction fn, void* data )
{
    dummyTest* test = dummyGetCurrentTest();

    assert(test->cleanupStackSize < DUMMY_MAX_CLEANUPS);
    dummyCleanup* cleanup = &test->cleanupStack[test->cleanupStackSize];

    cleanup->fn = fn;
    cleanup->data = data;

    test->cleanupStackSize++;
}

int dummyTestIsMarkedAsTodo()
{
    return dummyGetCurrentTest()->markedAsTodo;
}

const char* dummyGetTestTodoReason()
{
    const dummyTest* test = dummyGetCurrentTest();

    if(test->markedAsTodo)
    {
        assert(test->todoReason);
        const int length = strlen(test->todoReason);
        assert(length < DUMMY_MAX_MESSAGE_LENGTH);
        if(length > 0)
            return test->todoReason;
    }

    return NULL;
}

const char* dummyFormatV( const char* format, va_list args )
{
    static char buffer[DUMMY_MAX_MESSAGE_LENGTH];
    vsprintf(buffer, format, args);
    return buffer;
}

void dummyAbortTest( dummyTestAbortType type, const char* reason, ... )
{
    int errorCode = DUMMY_SANDBOX_GENERIC_ERROR;
    switch(type)
    {
        case DUMMY_FAIL_TEST:
            errorCode = DUMMY_SANDBOX_GENERIC_ERROR;
            break;

        case DUMMY_SKIP_TEST:
            errorCode = DUMMY_SANDBOX_SKIPPED;
            break;

        default:
            assert(!"Unknown abort type");
    }

    const char* formattedReason = NULL;
    if(reason)
    {
        va_list args;
        va_start(args, reason);
        formattedReason = dummyFormatV(reason, args);
        va_end(args);
    }
    else
    {
        assert(!"dummyAbortTest needs a reason!");
    }

    dummyAbortSandbox(errorCode, formattedReason);
}

void dummyMarkTestAsTodo( const char* reason, ... )
{
    dummyTest* test = dummyGetCurrentTest();
    test->markedAsTodo = true;
    if(reason)
    {
        va_list args;
        va_start(args, reason);
        strncpy(test->todoReason, dummyFormatV(reason, args), DUMMY_MAX_MESSAGE_LENGTH);
        va_end(args);
    }
}

void dummyLog( const char* message, ... )
{
    const dummyContext* ctx = dummyCurrentContext;
    assert(ctx);

    va_list args;
    va_start(args, message);
    const char* formattedMessage = dummyFormatV(message, args);
    va_end(args);

    ctx->reporter->log(ctx->reporter->context, formattedMessage);
}

void dummyPushAbortHandler( dummyAbortHandler handler, void* context )
{
    assert(AbortStackSize <= DUMMY_MAX_SANDBOX_DEPTH);
    AbortStackSize++;
    dummyAbortHandlerInfo* info = &AbortStack[AbortStackSize-1];

    info->handler = handler;
    info->context = context;
}

void dummyPopAbortHandler()
{
    assert(AbortStackSize >= 1);
    AbortStackSize--;
}

void dummyAbortSandbox( int errorCode, const char* reason )
{
    assert(AbortStackSize >= 1);
    dummyAbortHandlerInfo* info = &AbortStack[AbortStackSize-1];

    info->handler(info->context, errorCode, reason);
}
