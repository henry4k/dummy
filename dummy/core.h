#ifndef __DUMMY_CORE_H__
#define __DUMMY_CORE_H__

#ifdef __cplusplus
extern "C"
{
#endif

enum
{
    DUMMY_UNKNOWN_TEST_COUNT = -1
};

typedef enum
{
    DUMMY_TEST_PASSED,
    DUMMY_TEST_FAILED,
    DUMMY_TEST_SKIPPED
} dummyTestResult;

typedef enum
{
    DUMMY_FAIL_TEST,
    DUMMY_SKIP_TEST
} dummyTestAbortType;

enum
{
    DUMMY_SANDBOX_SUCEEDED = 0,
    DUMMY_SANDBOX_GENERIC_ERROR = 1
};

/**
 * Function that is called for a test.
 */
typedef void (*dummyTestFunction)();

typedef void (*dummyCleanupFunction)( void* data );

typedef struct
{
    void* context;

    /**
     * Runs the given function and catches errors.
     *
     * @param abortReason
     * If the protected call is aborted with a reason,
     * the string pointer is passed here.
     *
     * @return
     * Code that classifies the error.
     * If the function ras successfully it returns DUMMY_PROTECTED_CALL_SUCEEDED.
     * Custom error codes may be passed using dummySandbox::abort().
     */
    int (*run)( void* context, dummyTestFunction fn, const char** abortReason );

    /**
     * Aborts the current test.
     *
     * @param reason
     * May be `NULL`.
     *
     * @return
     * Doesn't return.
     */
    void (*abort)( void* context, int errorCode, const char* reason );
} dummySandbox;

typedef struct
{
    void* context;

    void (*began)( void* context );
    void (*completed)( void* context );

    void (*beganTest)( void* context );
    void (*completedTest)( void* context );

    void (*log)( void* context, const char* message );
} dummyReporter;


/**
 * Initializes the test context.
 *
 * This needs to be called before any calls to other test functions.
 * The initial context status is `DUMMY_INITIALIZING`.
 *
 * @param reporter
 * Reporter which will be used by the created context.
 */
void dummyInit( const dummySandbox* sandbox, const dummyReporter* reporter );

/**
 * Runs all added tests and destroys the current context.
 *
 * @return
 * Number of test tests, that failed.
 */
int dummyRunTests();

/**
 * Adds a test to the current context.
 *
 * @param name
 * May be used by the reporter to describe the test.
 *
 * @param fn
 * Is called when the test is being run.
 * See #dummyRunTests
 */
void dummyAddTest( const char* name, dummyTestFunction fn );

/**
 * Count of tests added to the current context.
 */
int dummyGetTestCount();

/**
 * Name of the active test in the current context.
 */
const char* dummyGetTestName();

/**
 * Id of the active test in the current context.
 */
int dummyGetTestNumber();

/**
 * TODO
 */
dummyTestResult dummyGetTestResult();

/**
 * Message describing why the current test aborted
 * or `NULL` if the test hasn't failed (yet).
 */
const char* dummyGetTestAbortReason();

/**
 * Adds a cleanup function to the current test.
 *
 * @param data
 * Data pointer that is passed to the cleanup function.
 */
void dummyAddCleanup( dummyCleanupFunction fn, void* data );

int dummyTestIsMarkedAsTodo();

/**
 * If set, it returns the reason why the current test has been marked as TODO.
 * May be `NULL` if no reason has been given.
 */
const char* dummyGetTestTodoReason();

/**
 * Aborts current test with the given reason.
 *
 * @param reason
 * Explains why the test was aborted -
 * this information is available to the reporter.
 *
 * @return
 * Doesn't return.
 */
void dummyAbortTest( dummyTestAbortType type, const char* reason, ... );

/**
 * Marks current test as TODO.
 *
 * @param reason
 * Explains why the test has been marked as TODO -
 * this information is available to the reporter.
 * May be `NULL` to signal an unknown reason.
 *
 * @note
 * Calling this function more than once in a test,
 * may overwrite the previous TODO reason.
 */
void dummyMarkTestAsTodo( const char* reason, ... );


/**
 * Reports a diagnostic message.
 */
void dummyLog( const char* message, ... );


#ifdef __cplusplus
}
#endif

#endif
