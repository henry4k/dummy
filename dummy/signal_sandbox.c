#include <stddef.h> // NULL
#include <setjmp.h>
#include <string.h> // memset, strncpy
#include <assert.h>


#include "signal_utils.h"
#include "signal_sandbox.h"


enum
{
    MAX_REASON_LENGTH = 512
};

typedef struct
{
    jmp_buf jumpBuffer;
    int errorCode;
    char abortReason[MAX_REASON_LENGTH];
} Context;


static Context* currentContext = NULL;


static void setAbortInformation( Context* context, int errorCode, const char* reason )
{
    context->errorCode = errorCode;
    if(reason)
        strncpy(context->abortReason, reason, MAX_REASON_LENGTH);
}

static void abortHandler( void* context_, int errorCode, const char* reason )
{
    Context* context = (Context*)context_;
    setAbortInformation(context, errorCode, reason);
    raise(SIGABRT);
}

static void signalHandler( int signal )
{
    assert(currentContext);

    // If setAbortInformation has not been called before:
    if(currentContext->errorCode == DUMMY_SANDBOX_SUCCEEDED)
    {
        setAbortInformation(
            currentContext,
            DUMMY_SANDBOX_GENERIC_ERROR,
            dummySignalToAbortReason(signal));
    }

    longjmp(currentContext->jumpBuffer, 1);
}

int dummySignalSandbox( dummySandboxableFunction fn, const char** abortReason )
{
    Context context;
    memset(&context, 0, sizeof(Context));
    context.errorCode = DUMMY_SANDBOX_SUCCEEDED;

    dummyPushAbortHandler(abortHandler, &context);
    dummyPushSignalHandler(signalHandler);
    Context* previousContext = currentContext;
    currentContext = &context;

    const int jumpResult = setjmp(context.jumpBuffer);
    if(jumpResult == 0)
        fn();

    currentContext = previousContext;
    dummyPopSignalHandler();
    dummyPopAbortHandler();

    // return results
    if(abortReason)
    {
        if(context.abortReason[0] == '\0')
            *abortReason = NULL;
        else
            *abortReason = context.abortReason;
    }
    return context.errorCode;
}
