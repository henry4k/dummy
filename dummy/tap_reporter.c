#include <stdio.h>
#include <assert.h>
#include "tap_reporter.h"

typedef struct
{
    FILE* file;
} Context;

static void began( void* ctx_ )
{
    const Context* ctx = (Context*)ctx_;
    fprintf(ctx->file, "1..%d\n", lamerGetTestCount());
}

static void completed( void* ctx_ )
{
}

static void beganTest( void* ctx_ )
{
}

static const char* lamerTestResultToTAPString( lamerTestResult result )
{
    switch(result)
    {
        case LAMER_TEST_PASSED:
            return "ok";
        case LAMER_TEST_FAILED:
            return "not ok";
        case LAMER_TEST_SKIPPED:
            return "ok";
        default:
            assert(!"Unknown result");
    }
}

static void completedTest( void* ctx_ )
{
    const Context* ctx = (Context*)ctx_;

    lamerTestResult result = lamerGetTestResult();
    const char* resultString = lamerTestResultToTAPString(result);
    const char* abortReason = lamerGetTestAbortReason();

    if(result == LAMER_TEST_FAILED && abortReason)
    {
        fprintf(ctx->file, "# %s\n", abortReason);
    }

    fprintf(ctx->file, "%s %d %s",
        resultString,
        lamerGetTestNumber()+1,
        lamerGetTestName()
    );

    if(lamerTestIsMarkedAsTodo())
    {
        fprintf(ctx->file, " # TODO");
        const char* todoReason = lamerGetTestTodoReason();
        if(todoReason)
            fprintf(ctx->file, " %s", todoReason);
    }
    else if(result == LAMER_TEST_SKIPPED)
    {
        fprintf(ctx->file, " # SKIP");
        if(abortReason)
            fprintf(ctx->file, " %s", abortReason);
    };

    fprintf(ctx->file, "\n");
}

static void diag( void* ctx_, const char* message )
{
    const Context* ctx = (Context*)ctx_;
    fprintf(ctx->file, "# %s\n", message);
}

const lamerReporter* lamerGetTAPReporter()
{
    static Context ctx;
    static lamerReporter reporter;

    ctx.file = stdout;

    reporter.context = &ctx;
    reporter.began = began;
    reporter.completed = completed;
    reporter.beganTest = beganTest;
    reporter.completedTest = completedTest;
    reporter.diag = diag;

    return &reporter;
}
