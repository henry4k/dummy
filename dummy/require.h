#ifndef __DUMMY_REQUIRE_H__
#define __DUMMY_REQUIRE_H__

#include "core.h"

#define dummyRequire( E ) dummyRequire_(#E, (E))

static inline void dummyRequire_( const char* expression, int result )
{
    if(!result)
        dummyAbortTest(DUMMY_FAIL_TEST, "%s failed", expression);
}

#endif
