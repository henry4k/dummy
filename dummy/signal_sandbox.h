#ifndef __DUMMY_SIGNAL_SANDBOX_H__
#define __DUMMY_SIGNAL_SANDBOX_H__

#include "core.h"

#ifdef __cplusplus
extern "C"
{
#endif

/**
 * Run `fn` in a sandbox, that catches signals.
 * @see dummySandbox
 */
int dummySignalSandbox( dummySandboxableFunction fn, const char** abortReason );

#ifdef __cplusplus
}
#endif

#endif
