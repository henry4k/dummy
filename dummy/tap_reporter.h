#ifndef __DUMMY_TAP_REPORTER_H__
#define __DUMMY_TAP_REPORTER_H__

#include <stdio.h> // FILE
#include "core.h"

#ifdef __cplusplus
extern "C"
{
#endif

/**
 * Reports using the 'test anything protocol' (TAP).
 *
 * Example:
 * @code
 * dummyGetTAPReporter(stdout)
 * @endcode
 *
 * @param file
 * The report is written to `file`.
 */
const dummyReporter* dummyGetTAPReporter( FILE* file );

#ifdef __cplusplus
}
#endif

#endif
