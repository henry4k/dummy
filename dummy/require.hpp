#ifndef __DUMMY_REQUIRE_HPP__
#define __DUMMY_REQUIRE_HPP__

#include "core.h"
#include "require.h"
#include <string.h> // strcmp, strlen
#include <sstream> // std::ostringstream

template<typename T>
bool _dummyEquals( T a, T b )
{
    return a == b;
}

template<char const *>
bool _dummyEquals( char const * a, char const * b )
{
    return strcmp(a, b) == 0;
}

#define dummyRequireEquals( expected, actual ) \
        dummyRequireEquals_((expected), (actual), \
                            #expected, #actual, \
                            __FILE__, __LINE__)
template<typename T>
void dummyRequireEquals_( T expected, T actual,
                          const char* expectedRepresentation,
                          const char* actualRepresentation,
                          const char* file, int line )
{
    if(!_dummyEquals(expected, actual))
    {
        using namespace std;
        ostringstream s;
        s << "In " << file << ":" << line << ":" << endl;
        s << "    " << actualRepresentation << " == " << expectedRepresentation << " failed" << endl;
        s << "was ";
        s.flags(ios::right);
        s.width(strlen(actualRepresentation));
        s << actual << "    ";
        s.flags(ios::left);
        s.width(strlen(expectedRepresentation));
        s << expected;
        dummyAbortTest(DUMMY_FAIL_TEST, s.str().c_str());
    }
}


#endif
