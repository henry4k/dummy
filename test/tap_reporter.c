#include <stdlib.h> // malloc, free
#include <stddef.h> // NULL
#include <stdio.h> // stdout
#include <assert.h>

#include <dummy/core.h>
#include <dummy/signal_sandbox.h>
#include <dummy/tap_reporter.h>
#include <dummy/require.h>


void PassingTest()
{
}

void AssertTest()
{
    assert(0);
}

void FailTest()
{
    dummyAbortTest(DUMMY_FAIL_TEST, "Fail reason");
}

void PassingRequireTest()
{
    dummyRequire(2+2 == 4);
}

void FailingRequireTest()
{
    dummyRequire(2+2 == 5);
}

void SkipTest()
{
    dummyAbortTest(DUMMY_SKIP_TEST, "Skip reason");
}

void TodoWithoutReasonTest()
{
    dummyMarkTestAsTodo(NULL);
}

void TodoWithReasonTest()
{
    dummyMarkTestAsTodo("TODO reason");
}

int main()
{
    dummyInit(dummyGetTAPReporter(stdout));
    dummyAddTest("PassingTest", dummySignalSandbox, PassingTest);
    dummyAddTest("AssertTest", dummySignalSandbox, AssertTest);
    dummyAddTest("FailTest", dummySignalSandbox, FailTest);
    dummyAddTest("PassingRequireTest", dummySignalSandbox, PassingRequireTest);
    dummyAddTest("FailingRequireTest", dummySignalSandbox, FailingRequireTest);
    dummyAddTest("SkipTest", dummySignalSandbox, SkipTest);
    dummyAddTest("TodoWithoutReasonTest", dummySignalSandbox, TodoWithoutReasonTest);
    dummyAddTest("TodoWithReasonTest", dummySignalSandbox, TodoWithReasonTest);
    dummyRunTests();
    return 0;
}
