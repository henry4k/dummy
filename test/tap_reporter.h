#ifndef __TEST_TAP_REPORTER_H__
#define __TEST_TAP_REPORTER_H__

#include "core.h"

#ifdef __cplusplus
extern "C"
{
#endif

/**
 * Reports using the 'test anything protocol' (TAP).
 */
const testReporter* testGetTAPReporter();

#ifdef __cplusplus
}
#endif

#endif
