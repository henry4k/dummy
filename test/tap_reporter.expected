1..8
ok 1 PassingTest
# SIGABRT: Abnormal termination
not ok 2 AssertTest
# Fail reason
not ok 3 FailTest
ok 4 PassingRequireTest
# 2+2 == 5 failed
not ok 5 FailingRequireTest
ok 6 SkipTest # SKIP Skip reason
ok 7 TodoWithoutReasonTest # TODO
ok 8 TodoWithReasonTest # TODO TODO reason
